orchagent/pfcactionhandler.cpp

```diff
@@ -133,6 +133,86 @@ void PfcWdActionHandler::updateWdCounters(const string& queueIdStr, const PfcWdQ
     m_countersTable->set(queueIdStr, resultFvValues);
 }
 
+PfcWdAclHandler::PfcWdAclHandler(sai_object_id_t port, sai_object_id_t queue,
+        uint8_t queueId, shared_ptr<Table> countersTable):
+    PfcWdActionHandler(port, queue, queueId, countersTable)
+{
+    acl_table_type_t table_type = ACL_TABLE_L3;
+
+    // There is one handler instance per queue ID
+    string queuestr = to_string(queueId);
+    m_strTable = "Table_PfcWdAclHandler_" + queuestr;
+    m_strRule = "Rule_PfcWdAclHandler_" + queuestr;
+
+    auto found = m_aclTables.find(m_strTable);
+    if (found == m_aclTables.end())
+    {
+        // First time of handling PFC for this queue, create ACL table, and bind
+        createPfcAclTable(port);
+        shared_ptr<AclRuleL3> newRule = make_shared<AclRuleL3>(gAclOrch, m_strRule, m_strTable, table_type);
+        createPfcAclRule(newRule, queueId);
+    }
+    else
+    {
+        // Otherwise just bind ACL table with the port
+        found->second.bind(port);
+    }
+}
+
+PfcWdAclHandler::~PfcWdAclHandler(void)
+{
+    SWSS_LOG_ENTER();
+
+    auto found = m_aclTables.find(m_strTable);
+    found->second.unbind(getPort());
+}
+
+void PfcWdAclHandler::clear()
+{
+    for (auto& tablepair: m_aclTables)
+    {
+        auto& table = tablepair.second;
+        gAclOrch->removeAclTable(table.getId());
+    }
+}
+
+void PfcWdAclHandler::createPfcAclTable(sai_object_id_t port)
+{
+    SWSS_LOG_ENTER();
+
+    auto inserted = m_aclTables.emplace(piecewise_construct,
+        std::forward_as_tuple(m_strTable),
+        std::forward_as_tuple());
+
+    assert(inserted.second);
+    AclTable& aclTable = inserted.first->second;
+    aclTable.type = ACL_TABLE_L3;
+    aclTable.link(port);
+    aclTable.id = m_strTable;
+    gAclOrch->addAclTable(aclTable, m_strTable);
+}
+
+void PfcWdAclHandler::createPfcAclRule(shared_ptr<AclRuleL3> rule, uint8_t queueId)
+{
+    string attr_name, attr_value;
+
+    attr_name = "RULE_PRIORITY";
+    attr_value = "999";
+    rule->validateAddPriority(attr_name, attr_value);
+
+    attr_name = "MATCH_TC";
+    attr_value = to_string(queueId);
+    rule->validateAddMatch(attr_name, attr_value);
+
+    attr_name = ACTION_PACKET_ACTION;
+    attr_value = PACKET_ACTION_DROP;
+    rule->validateAddAction(attr_name, attr_value);
+
+    gAclOrch->addAclRule(rule, m_strTable);
```

Comment:
When is the rule removed?

Reply:
Not clear in requirement, so I assume them be long living. I created a function clear() to remove all the tables and rules, and it could be called if we make different decision.

Let me know if you have any other idea?